File: src-tauri/src/gcode_conversion/stroke_to_gcode.rs
use super::coordinate_converter::CoordinateConverter;
use super::types::{PlotterMove, PlotterStroke};

pub fn stroke_to_gcode(stroke: &PlotterStroke, conv: &CoordinateConverter) -> String {
    let mut gcode = String::new();

    let pen_down = if stroke.reversed {
        conv.doc_xy_to_gcode_xy(stroke.end.0, stroke.end.1)
    } else {
        conv.doc_xy_to_gcode_xy(stroke.start.0, stroke.start.1)
    };
    gcode.push_str(&format!("G0 X{:.3} Y{:.3}\n", pen_down.0, pen_down.1));
    gcode.push_str("M3\n"); // pen down

    let moves_iter: Box<dyn Iterator<Item = &PlotterMove>> = if stroke.reversed {
        Box::new(stroke.moves.iter().rev())
    } else {
        Box::new(stroke.moves.iter())
    };

    for m in moves_iter {
        match m {
            PlotterMove::Line { x1, y1, x2, y2 } => {
                // When reversed, travel toward (x1,y1); otherwise toward (x2,y2).
                let (tx, ty) = if stroke.reversed {
                    conv.doc_xy_to_gcode_xy(*x1, *y1)
                } else {
                    conv.doc_xy_to_gcode_xy(*x2, *y2)
                };
                gcode.push_str(&format!("G1 X{:.3} Y{:.3}\n", tx, ty));
            }
            PlotterMove::Arc { x1, y1, cx, cy, x2, y2, clockwise } => {
                // Reversal swaps travel direction (flips winding); Y-flip also flips winding.
                // Two flips cancel out: reversed arcs use the same G2/G3 as the original.
                // Non-reversed arcs: Y-flip alone inverts, so CW doc → G3 gcode.
                let (from_x, from_y, to_x, to_y) = if stroke.reversed {
                    (*x2, *y2, *x1, *y1)
                } else {
                    (*x1, *y1, *x2, *y2)
                };
                let (gto_x, gto_y)     = conv.doc_xy_to_gcode_xy(to_x, to_y);
                let (gcx, gcy)         = conv.doc_xy_to_gcode_xy(*cx, *cy);
                let (gfrom_x, gfrom_y) = conv.doc_xy_to_gcode_xy(from_x, from_y);
                // Y-flip inverts winding; reversal inverts again — net: reversed = original winding.
                let effective_cw = if stroke.reversed { *clockwise } else { !clockwise };
                let command = if effective_cw { "G2" } else { "G3" };
                gcode.push_str(&format!(
                    "{} X{:.3} Y{:.3} I{:.3} J{:.3}\n",
                    command, gto_x, gto_y, gcx - gfrom_x, gcy - gfrom_y,
                ));
            }
            PlotterMove::QuadraticBezier { x1, y1, cx, cy, x2, y2 } => {
                // Reversed: travel from (x2,y2) toward (x1,y1); control point is unchanged.
                let (to_x, to_y) = if stroke.reversed { (*x1, *y1) } else { (*x2, *y2) };
                let (gto_x, gto_y) = conv.doc_xy_to_gcode_xy(to_x, to_y);
                let (gcx, gcy)     = conv.doc_xy_to_gcode_xy(*cx, *cy);
                gcode.push_str(&format!("G5.1 X{:.3} Y{:.3} CX{:.3} CY{:.3}\n", gto_x, gto_y, gcx, gcy));
            }
            PlotterMove::CubicBezier { x1, y1, cx1, cy1, cx2, cy2, x2, y2 } => {
                // Reversed: travel from (x2,y2) toward (x1,y1); swap the two control points.
                let (to_x, to_y, gcx1, gcy1, gcx2, gcy2) = if stroke.reversed {
                    let (gto_x, gto_y) = conv.doc_xy_to_gcode_xy(*x1, *y1);
                    let (gc1x, gc1y)   = conv.doc_xy_to_gcode_xy(*cx2, *cy2);
                    let (gc2x, gc2y)   = conv.doc_xy_to_gcode_xy(*cx1, *cy1);
                    (gto_x, gto_y, gc1x, gc1y, gc2x, gc2y)
                } else {
                    let (gto_x, gto_y) = conv.doc_xy_to_gcode_xy(*x2, *y2);
                    let (gc1x, gc1y)   = conv.doc_xy_to_gcode_xy(*cx1, *cy1);
                    let (gc2x, gc2y)   = conv.doc_xy_to_gcode_xy(*cx2, *cy2);
                    (gto_x, gto_y, gc1x, gc1y, gc2x, gc2y)
                };
                gcode.push_str(&format!(
                    "G5 X{:.3} Y{:.3} CX1{:.3} CY1{:.3} CX2{:.3} CY2{:.3}\n",
                    to_x, to_y, gcx1, gcy1, gcx2, gcy2,
                ));
            }
        }
    }
    gcode.push_str("M5\n"); // pen up
    gcode
}

File: src/features/document/gcode/strokeToGcode.ts
import type { PlotterMove, GStroke } from "./types";
import { type Converter, docToGcode } from "./converter";

function moveToGcode(m: PlotterMove, reversed: boolean, conv: Converter): string {
  switch (m.type) {
    case "Line": {
      const [tx, ty] = reversed ? docToGcode(m.x1, m.y1, conv) : docToGcode(m.x2, m.y2, conv);
      return `G1 X${tx.toFixed(3)} Y${ty.toFixed(3)}\n`;
    }

    case "Arc": {
      const [fromX, fromY, toX, toY] = reversed
        ? [m.x2, m.y2, m.x1, m.y1]
        : [m.x1, m.y1, m.x2, m.y2];
      const [gToX, gToY]     = docToGcode(toX, toY, conv);
      const [gCx,  gCy ]     = docToGcode(m.cx, m.cy, conv);
      const [gFrX, gFrY]     = docToGcode(fromX, fromY, conv);
      // Y-flip inverts winding; reversal inverts again — net: reversed = original winding.
      const effectiveCw = reversed ? m.clockwise : !m.clockwise;
      const cmd = effectiveCw ? "G2" : "G3";
      return `${cmd} X${gToX.toFixed(3)} Y${gToY.toFixed(3)} I${(gCx - gFrX).toFixed(3)} J${(gCy - gFrY).toFixed(3)}\n`;
    }

    case "QuadBezier": {
      const [toX, toY]   = reversed ? [m.x1, m.y1] : [m.x2, m.y2];
      const [gToX, gToY] = docToGcode(toX, toY, conv);
      const [gCx,  gCy ] = docToGcode(m.cx, m.cy, conv);
      return `G5.1 X${gToX.toFixed(3)} Y${gToY.toFixed(3)} CX${gCx.toFixed(3)} CY${gCy.toFixed(3)}\n`;
    }

    case "CubicBezier": {
      // When reversed, swap the two control points.
      const [toX, toY, cx1, cy1, cx2, cy2] = reversed
        ? [m.x1, m.y1, m.cx2, m.cy2, m.cx1, m.cy1]
        : [m.x2, m.y2, m.cx1, m.cy1, m.cx2, m.cy2];
      const [gToX, gToY] = docToGcode(toX, toY, conv);
      const [gc1X, gc1Y] = docToGcode(cx1, cy1, conv);
      const [gc2X, gc2Y] = docToGcode(cx2, cy2, conv);
      return `G5 X${gToX.toFixed(3)} Y${gToY.toFixed(3)} CX1${gc1X.toFixed(3)} CY1${gc1Y.toFixed(3)} CX2${gc2X.toFixed(3)} CY2${gc2Y.toFixed(3)}\n`;
    }
  }
}

export function strokeToGcode(stroke: GStroke, conv: Converter): string {
  let gcode = "";

  if (stroke.kind === "open") {
    const origin = stroke.reversed ? stroke.end : stroke.start;
    const [px, py] = docToGcode(origin[0], origin[1], conv);
    gcode += `G0 X${px.toFixed(3)} Y${py.toFixed(3)}\n`;
    gcode += "M3\n"; // pen down
    const moves = stroke.reversed ? [...stroke.moves].reverse() : stroke.moves;
    for (const m of moves) gcode += moveToGcode(m, stroke.reversed, conv);
    gcode += "M5\n"; // pen up

  } else {
    const joint = stroke.joints[stroke.startIndex];
    const [px, py] = docToGcode(joint[0], joint[1], conv);
    gcode += `G0 X${px.toFixed(3)} Y${py.toFixed(3)}\n`;
    gcode += "M3\n"; // pen down
    const n = stroke.moves.length;
    for (let i = 0; i < n; i++) {
      // Rotate moves to start at startIndex; loops are never reversed.
      gcode += moveToGcode(stroke.moves[(stroke.startIndex + i) % n], false, conv);
    }
    gcode += "M5\n"; // pen up
  }

  return gcode;
}
